librad/src/git/local/url.rs

```diff
@@ -0,0 +1,107 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    fmt::{self, Display},
+    str::FromStr,
+};
+
+use thiserror::Error;
+
+use crate::{
+    hash::{self, Hash},
+    uri::{self, RadUrn},
+};
+
+#[derive(Clone)]
+pub struct LocalUrl {
+    repo: Hash,
+}
+
+impl LocalUrl {
+    pub fn repo(&self) -> &Hash {
+        &self.repo
+    }
+}
+
+impl Display for LocalUrl {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}://{}.git", super::URL_SCHEME, self.repo)
+    }
+}
+
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum ParseError {
+    #[error("Invalid scheme: {0}")]
+    InvalidScheme(String),
+
+    #[error("Cannot-be-a-base URL")]
+    CannotBeABase,
+
+    #[error("Malformed URL")]
```

Comment:
Why not transparent here?

Reply:
Because I didn't think it's too interesting -- the URL is supposed to be constructed programmatically via `LocalUrl`. Feel free to change.